include/pybind11/cast.h

```diff
@@ -1524,6 +1524,37 @@ struct holder_helper {
     static auto get(const T &p) -> decltype(p.get()) { return p.get(); }
 };
 
+/// Another helper class for holders that helps construct derivative holders from
+/// the original holder
+template <typename T>
+struct holder_retriever {
+    static auto get_derivative_holder(const value_and_holder &v_h) -> decltype(v_h.template holder<T>()) {
+        return v_h.template holder<T>();
+    }
+};
+
+template <typename T>
+struct holder_retriever<std::shared_ptr<T>> {
+    struct shared_ptr_deleter {
```

Comment:
> The delete expression must be well-formed, have well-defined behavior and not throw any exceptions. 

... I feel like this could throw? Alternatives?

Reply:
This is done in the std::function destructor too, so not going to deal with it for now.